beacon/src/entropy.rs

```diff
@@ -0,0 +1,103 @@
+use rand::{rngs::OsRng, RngCore};
+use serde::{Deserialize, Serialize};
+use sha2::Digest;
+use std::time::{Duration, SystemTime, UNIX_EPOCH};
+
+pub const LOCAL_ENTROPY_SIZE: usize = 4;
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
+pub struct Entropy {
+    #[serde(default = "default_version")]
+    pub version: u32,
+    pub timestamp: i64,
+    #[serde(with = "serde_base64")]
+    pub data: Vec<u8>,
+}
+
+impl Entropy {
+    pub fn local() -> Self {
+        let mut local_entropy = [0u8; LOCAL_ENTROPY_SIZE];
+        OsRng.fill_bytes(&mut local_entropy);
+        let version = default_version();
+        let timestamp = SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .unwrap_or_else(|_| Duration::from_secs(0))
+            .as_secs() as i64;
+        Self {
+            version,
+            timestamp,
+            data: local_entropy.to_vec(),
```

Comment:
Might as well keep the Vec and skip the intermediate stack allocation

```suggestion
        let mut local_entropy = vec![0u8; LOCAL_ENTROPY_SIZE];
        OsRng.fill_bytes(&mut local_entropy);
        let version = default_version();
        let timestamp = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_else(|_| Duration::from_secs(0))
            .as_secs() as i64;
        Self {
            version,
            timestamp,
            data: local_entropy,
```